### DISCUSSION TITLE
Plugin API

### DISCUSSION THREAD

**Discussion by Akryum** [histoire-dev/histoire#240] [Ideas]
Currently the Plugin API is barebone (and undocumented :sweat_smile: ) and doesn't allow customizing the books much. This thread is about discussing how we want to expand this API to allow more customization and flexibility.

What we can do with the Plugin API [right now](https://github.com/histoire-dev/histoire/blob/ad3b3ed0e20deacc3775dae44cf04037fcf961a7/packages/histoire/src/node/plugin.ts):
- Customize the Histoire config
- Hook on Histoire config resolved
- Hook on dev server start
- Hook on book build (ex: [percy snapshots](https://github.com/histoire-dev/histoire/blob/ad3b3ed0e20deacc3775dae44cf04037fcf961a7/packages/histoire-plugin-percy/src/index.ts))
- Define support plugins (ex: [vue](https://github.com/histoire-dev/histoire/blob/ad3b3ed0e20deacc3775dae44cf04037fcf961a7/packages/histoire-plugin-vue/src/index.ts), [svelte](https://github.com/histoire-dev/histoire/blob/ad3b3ed0e20deacc3775dae44cf04037fcf961a7/packages/histoire-plugin-svelte/src/index.ts), etc.)
- Automatically generate stories in a temp folder (ex: [tailwind design tokens](https://github.com/histoire-dev/histoire/blob/ad3b3ed0e20deacc3775dae44cf04037fcf961a7/packages/histoire/src/node/builtin-plugins/tailwind-tokens.ts))
So nothing about modifying the generated web app yet.

Wishlist:
- [widget] Be able to render a component to be used in the client (aka widget)
- [story-tab] Add a tab to the story pane (ex: #85, #154) and render a widget
- [main-menu] Add buttons to the main toolbar and render a menu with click callbacks (or external links) or a widget
- [story-menu] Add buttons to the story toolbar (next to background color and responsive pickers) and render a widget
- ???

Have any idea or questions, feel free to post them here!

**Comment by agritheory** [comment]
We're interested in developing a widget-like plugin that would allow for switching app-level themes in the story. Use case for this is the normal dark/light mode stuff, but also being able to apply client's themes to the components under development. I imagine the API for our use case to look something like. this:

```js
import { defineConfig } from 'histoire'
import { ThemeSwitcherPlugin, pathResolver } from 'my-histoire-plugin'

export default defineConfig({
  plugins: [
    ThemeSwitcherPlugin({
        themes: [
           {title: 'Dark', path: 'path/to/style.css'},
           {title: 'Client A', path: 'path/to/cleint_a_style.css'},
           {title: 'Functional', path: () => {return pathResolver('special_case.css')}},
        ]
    }),
  ],
})
```
These would be shown in a `<select>`, ideally here:
![image](https://user-images.githubusercontent.com/18033666/184495847-4ca9278d-8745-4a0a-b2ed-efe903b6fed8.png)

Just providing this as a concept because I don't think we'll be the only ones with this kind of requirement.

**Reply by Akryum** [reply]
Added [story-menu]

**Reply by Akryum** [reply]
I guess it would be nice to have menus use a query variable in the URL

**Reply by agritheory** [reply]
This is a use case where I think that would be really valuable. Is [story-menu] another feature?

**Reply by Akryum** [reply]
See the original post :D

**Comment by alvarosabu** [comment]
Nice thread, some ideas from the a11y one

- Possibility to add extra tabs where the docs event tabs are, probably think how should it behave if there are too many tabs.
- Lifecycle hooks 
- Access stories data
- How to render on sidebars.

**Reply by Akryum** [reply]
>  probably think how should it behave if there are too many tab

It already puts the overflowing tabs into a dropdown menu 😉

**Comment by MichaelTessier** [comment]
Hello, have you planned [story-tab] feature ?

**Reply by MichaelTessier** [reply]
up @Akryum  please, any idea for the story-tab feature ?
